wrapper/src/test/java/software/amazon/jdbc/plugin/cache/CacheMonitorTest.java

```diff
@@ -0,0 +1,544 @@
+    assertEquals(CacheMonitor.HealthState.HEALTHY, cluster.roHealthState);
+    verify(mockErrorCounter, times(1)).inc();
+    verify(mockStateTransitionCounter, times(1)).inc();
+
+    // Test 2: Non-recoverable error doesn't change state
+    CacheMonitor.reportError("localhost:6379", "localhost:6380", true,
+        new RuntimeException("Serialization failed"), "SET");
+    assertEquals(CacheMonitor.HealthState.SUSPECT, cluster.rwHealthState);
+    verify(mockErrorCounter, times(2)).inc();
+    verify(mockStateTransitionCounter, times(1)).inc();
+
+    // Test 3: RO endpoint error transitions independently
+    CacheMonitor.reportError("localhost:6379", "localhost:6380", false,
+        new RedisCommandExecutionException("READONLY"), "SET");
+    assertEquals(CacheMonitor.HealthState.SUSPECT, cluster.rwHealthState);
+    assertEquals(CacheMonitor.HealthState.SUSPECT, cluster.roHealthState);
+    verify(mockErrorCounter, times(3)).inc();
+    verify(mockStateTransitionCounter, times(2)).inc();
+
+    // Test 4: Unregistered cluster logs warning without metrics
+    CacheMonitor.reportError("localhost:9999", null, true,
+        new RedisConnectionException("Connection refused"), "SET");
+    verify(mockErrorCounter, times(3)).inc();
+    verify(mockStateTransitionCounter, times(2)).inc();
+  }
+
+  @Test
+  void testInFlightSize_MemoryPressureScenarios() throws Exception {
+    props.setProperty("cacheInFlightWriteSizeLimitBytes", "1000");
+    registerClusterWithTelemetry("localhost:6379", null);
+    CacheMonitor.ClusterHealthState cluster = getCluster("localhost:6379", null);
+
+    // Test 1: Below limit remains healthy
+    CacheMonitor.incrementInFlightSizeStatic("localhost:6379", null, 500);
+    assertEquals(CacheMonitor.HealthState.HEALTHY, cluster.rwHealthState);
+    assertEquals(500L, cluster.inFlightWriteSizeBytes.get());
+
+    // Test 2: Exceeds limit transitions to degraded
+    CacheMonitor.incrementInFlightSizeStatic("localhost:6379", null, 1000);
+    assertEquals(CacheMonitor.HealthState.DEGRADED, cluster.rwHealthState);
+    assertEquals(1500L, cluster.inFlightWriteSizeBytes.get());
+    verify(mockStateTransitionCounter, times(1)).inc();
+
+    // Test 3: Decrement reduces size
+    CacheMonitor.decrementInFlightSizeStatic("localhost:6379", null, 900);
+    assertEquals(600L, cluster.inFlightWriteSizeBytes.get());
+
+    // Test 4: Decrement never goes negative
+    CacheMonitor.decrementInFlightSizeStatic("localhost:6379", null, 1000);
+    assertEquals(0L, cluster.inFlightWriteSizeBytes.get());
+  }
+
+  @Test
+  void testGetClusterState() throws Exception {
+    // Test 1: Single endpoint state transitions
+    registerCluster("localhost:6379", null);
+    CacheMonitor.ClusterHealthState cluster = getCluster("localhost:6379", null);
+    assertEquals(CacheMonitor.HealthState.HEALTHY,
+        CacheMonitor.getClusterState("localhost:6379", null));
+
+    cluster.rwHealthState = CacheMonitor.HealthState.SUSPECT;
```

Comment:
Not sure if that is the way you want to update the cluster health state via directly accessing the fields in `cluster` struct.